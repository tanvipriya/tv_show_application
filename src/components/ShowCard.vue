<template>
  <div class="show-card">
    <img :src="show.image?.medium || placeholder" :alt="show.name" />
    <h3 class="show-name">{{ show.name }}</h3>
    <p class="show-rating">‚≠ê {{ show.rating?.average || "N/A" }}</p>
    <p class="show-year">{{ show.premiered?.slice(0, 4) || "N/A" }}</p>
  </div>
</template>

<script setup>
defineProps({
  show: Object,
});

const placeholder = "https://via.placeholder.com/150x225?text=No+Image";
</script>

<style scoped>
.show-card {
  min-width: 150px;
  margin: 0 8px;
  text-align: center;
  flex-shrink: 0;
}
.show-card img {
  width: 100%;
  border-radius: 8px;
  object-fit: cover;
}
.show-name {
  font-size: 0.9rem;
  margin: 4px 0 2px 0;
}
.show-rating,
.show-year {
  font-size: 0.8rem;
  margin: 1px 0;
}
</style>
