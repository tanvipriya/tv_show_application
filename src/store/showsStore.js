import { defineStore } from 'pinia';
import axios from 'axios';

export const useShowsStore = defineStore('shows', {
    state: () => ({
        showsByGenre: {}, // { drama: [...], comedy: [...] }
        loading: false,
        error: null
    }),
    actions: {
        async fetchShowsByGenre(genre) {
            this.loading = true;
            this.error = null;
            try {
                // Fetch first 2 pages (TVMaze has 250 shows per page)
                const pages = [0, 1];
                let allShows = [];

                for (const page of pages) {
                    const res = await axios.get(`https://api.tvmaze.com/shows?page=${page}`);
                    allShows = allShows.concat(res.data);
                }

                // Filter by genre
                const filtered = allShows.filter((show) => show.genres.includes(genre));

                // Sort by rating descending (null rating goes last)
                filtered.sort((a, b) => (b.rating.average || 0) - (a.rating.average || 0));

                this.showsByGenre[genre] = filtered;
            } catch (err) {
                this.error = err.message || 'Failed to fetch shows';
            } finally {
                this.loading = false;
            }
        }
    }
});
