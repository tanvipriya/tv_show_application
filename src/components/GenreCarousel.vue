<template>
  <div class="genre-carousel">
    <div class="carousel-header">
      <h2>{{ genre }}</h2>
      <router-link
        :to="{ name: 'GenreView', params: { genre } }"
        class="see-all"
      >
        See All
      </router-link>
    </div>

    <div class="carousel-container">
      <router-link
        v-for="show in shows"
        :key="show.id"
        :to="{ name: 'ShowDetail', params: { id: show.id } }"
        class="show-link"
      >
        <ShowCard :show="show" />
      </router-link>
    </div>

    <p v-if="loading">Loading...</p>
    <p v-if="error" class="error">{{ error }}</p>
  </div>
</template>

<script setup>
import { onMounted, computed } from "vue";
import { useShowsStore } from "@/store/showsStore";
import ShowCard from "./ShowCard.vue";

const props = defineProps({
  genre: String,
});

const store = useShowsStore();

// Compute shows for this genre
const shows = computed(() => store.showsByGenre[props.genre] || []);
const loading = computed(() => store.loading);
const error = computed(() => store.error);

onMounted(async () => {
  // Load first batch if not loaded yet
  if (!store.showsByGenre[props.genre]) {
    await store.fetchNextPage(props.genre); // <--- updated function
  }
});
</script>

<style scoped>
.genre-carousel {
  margin-bottom: 24px;
}

.carousel-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 8px;
}

.carousel-container {
  display: flex;
  overflow-x: auto;
  scroll-behavior: smooth;
  gap: 12px;
  padding-bottom: 8px;
}

.carousel-container::-webkit-scrollbar {
  height: 8px;
}

.carousel-container::-webkit-scrollbar-thumb {
  background: #888;
  border-radius: 4px;
}

.show-link {
  text-decoration: none;
  color: inherit;
}

.see-all {
  font-size: 0.85rem;
  color: #004c4d;
  text-decoration: none;
}

.error {
  color: red;
}
</style>
