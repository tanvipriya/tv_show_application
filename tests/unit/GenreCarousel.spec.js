import { createPinia, setActivePinia } from 'pinia';
import { useShowsStore } from '@/store/showsStore';
import { mount } from '@vue/test-utils';
import GenreCarousel from '@/components/GenreCarousel.vue';
import ShowCard from '@/components/ShowCard.vue';

describe('GenreCarousel Component', () => {
    let pinia;
    let store;

    beforeEach(() => {
        pinia = createPinia();
        setActivePinia(pinia);
        store = useShowsStore();

        // Default store state
        store.showsByGenre = {
            Drama: [
                { id: 1, name: 'Breaking Bad', rating: { average: 9.5 }, image: { medium: '' } },
                { id: 2, name: 'Game of Thrones', rating: { average: 9.3 }, image: { medium: '' } },
            ],
        };
        store.loading = false;
        store.error = null;
        store.fetchNextPage = jest.fn();
    });

    // Helper to mount component with proper RouterLink stub
    const mountCarousel = (props) => {
        return mount(GenreCarousel, {
            global: {
                plugins: [pinia],
                stubs: {
                    RouterLink: {
                        template: '<div><slot /></div>', // render children to allow ShowCard to mount
                    },
                },
            },
            props,
        });
    };

    it('renders shows correctly and matches snapshot', () => {
        const wrapper = mountCarousel({ genre: 'Drama' });

        const showLinks = wrapper.findAll('.show-link');
        expect(showLinks.length).toBe(2);

        expect(wrapper.html()).toMatchSnapshot('GenreCarousel - normal shows');
    });

    it('renders loading state correctly', () => {
        store.loading = true;

        const wrapper = mountCarousel({ genre: 'Drama' });

        const skeleton = wrapper.find('.skeleton-wrapper');
        expect(skeleton.exists()).toBe(true);
        expect(skeleton.findAll('.show-skeleton').length).toBe(5);

        expect(wrapper.html()).toMatchSnapshot('GenreCarousel - loading state');
    });

    it('renders empty state correctly when no shows are available', () => {
        store.showsByGenre = { Drama: [] };
        store.loading = false;
        store.error = null;

        const wrapper = mountCarousel({ genre: 'Drama' });

        const emptyEl = wrapper.find('.empty');
        expect(emptyEl.exists()).toBe(true);
        expect(emptyEl.text()).toBe('No shows available');

        expect(wrapper.html()).toMatchSnapshot('GenreCarousel - empty state');
    });

    it('renders error state correctly', () => {
        store.error = 'Failed to fetch shows';

        const wrapper = mountCarousel({ genre: 'Drama' });

        const errorEl = wrapper.find('.error');
        expect(errorEl.exists()).toBe(true);
        expect(errorEl.text()).toBe('Failed to fetch shows');

        expect(wrapper.html()).toMatchSnapshot('GenreCarousel - error state');
    });

    it('handles click safely on first show', async () => {
        const wrapper = mountCarousel({ genre: 'Drama' });

        const showLinks = wrapper.findAll('.show-link');
        expect(showLinks.length).toBe(2);

        await showLinks[0].trigger('click');
    });

    it('calls fetchNextPage when mounted if genre not loaded', () => {
        store.showsByGenre = {}; // simulate genre not loaded

        mountCarousel({ genre: 'Drama' });

        expect(store.fetchNextPage).toHaveBeenCalledWith('Drama');
    });

    it('does not call fetchNextPage if genre already loaded', () => {
        mountCarousel({ genre: 'Drama' });

        expect(store.fetchNextPage).not.toHaveBeenCalled();
    });

    it('sorts shows by rating descending', () => {
        store.showsByGenre = {
            Drama: [
                { id: 1, name: 'Low Rated Show', rating: { average: 2 } },
                { id: 2, name: 'High Rated Show', rating: { average: 9 } },
            ],
        };
        store.loading = false;

        const wrapper = mountCarousel({ genre: 'Drama' });

        const showTexts = wrapper.findAllComponents(ShowCard).map(s => s.props('show').name);
        expect(showTexts).toEqual(['High Rated Show', 'Low Rated Show']);
        expect(wrapper.html()).toMatchSnapshot('GenreCarousel - sorted shows');
    });

    it('handles shows without rating gracefully', () => {
        store.showsByGenre = {
            Drama: [
                { id: 1, name: 'No Rating Show' }, // missing rating
                { id: 2, name: 'Rated Show', rating: { average: 5 } },
            ],
        };
        store.loading = false;

        const wrapper = mountCarousel({ genre: 'Drama' });

        const showTexts = wrapper.findAllComponents(ShowCard).map(s => s.props('show').name);
        expect(showTexts).toEqual(['Rated Show', 'No Rating Show']);
        expect(wrapper.html()).toMatchSnapshot('GenreCarousel - missing rating shows');
    });
});
