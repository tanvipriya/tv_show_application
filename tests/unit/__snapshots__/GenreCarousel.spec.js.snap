// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreCarousel Component handles shows without rating gracefully: GenreCarousel - missing rating shows 1`] = `
<div class="genre-carousel">
  <div class="carousel-header">
    <h2>Drama</h2>
    <div to="[object Object]" class="see-all"> See All </div>
  </div>
  <div class="carousel-container">
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Rated Show</h3>
          <p>⭐ 5</p>
          <p>—</p>
        </div>
      </div>
    </div>
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>No Rating Show</h3>
          <p>⭐ N/A</p>
          <p>—</p>
        </div>
      </div>
    </div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>
`;

exports[`GenreCarousel Component renders empty state correctly when no shows are available: GenreCarousel - empty state 1`] = `
<div class="genre-carousel">
  <div class="carousel-header">
    <h2>Drama</h2>
    <div to="[object Object]" class="see-all"> See All </div>
  </div>
  <div class="carousel-container">
    <div class="empty"> No shows available </div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>
`;

exports[`GenreCarousel Component renders error state correctly: GenreCarousel - error state 1`] = `
<div class="genre-carousel">
  <div class="carousel-header">
    <h2>Drama</h2>
    <div to="[object Object]" class="see-all"> See All </div>
  </div>
  <div class="carousel-container">
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Breaking Bad</h3>
          <p>⭐ 9.5</p>
          <p>—</p>
        </div>
      </div>
    </div>
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Game of Thrones</h3>
          <p>⭐ 9.3</p>
          <p>—</p>
        </div>
      </div>
    </div>
  </div>
  <!--v-if-->
  <p class="error">Failed to fetch shows</p>
</div>
`;

exports[`GenreCarousel Component renders loading state correctly: GenreCarousel - loading state 1`] = `
<div class="genre-carousel">
  <div class="carousel-header">
    <h2>Drama</h2>
    <div to="[object Object]" class="see-all"> See All </div>
  </div>
  <div class="carousel-container">
    <div class="skeleton-wrapper">
      <div class="show-skeleton"></div>
      <div class="show-skeleton"></div>
      <div class="show-skeleton"></div>
      <div class="show-skeleton"></div>
      <div class="show-skeleton"></div>
    </div>
  </div>
  <p>Loading...</p>
  <!--v-if-->
</div>
`;

exports[`GenreCarousel Component renders shows correctly and matches snapshot: GenreCarousel - normal shows 1`] = `
<div class="genre-carousel">
  <div class="carousel-header">
    <h2>Drama</h2>
    <div to="[object Object]" class="see-all"> See All </div>
  </div>
  <div class="carousel-container">
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Breaking Bad</h3>
          <p>⭐ 9.5</p>
          <p>—</p>
        </div>
      </div>
    </div>
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Game of Thrones</h3>
          <p>⭐ 9.3</p>
          <p>—</p>
        </div>
      </div>
    </div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>
`;

exports[`GenreCarousel Component sorts shows by rating descending: GenreCarousel - sorted shows 1`] = `
<div class="genre-carousel">
  <div class="carousel-header">
    <h2>Drama</h2>
    <div to="[object Object]" class="see-all"> See All </div>
  </div>
  <div class="carousel-container">
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>High Rated Show</h3>
          <p>⭐ 9</p>
          <p>—</p>
        </div>
      </div>
    </div>
    <div to="[object Object]" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Low Rated Show</h3>
          <p>⭐ 2</p>
          <p>—</p>
        </div>
      </div>
    </div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>
`;
